x/blob/README.md

```diff
@@ -2,49 +2,192 @@
+`MsgPayForBlobs` pays for the inclusion of a blob in the block and consists of the
+following fields:
+
+- signer: bech32 encoded signer address
+- namespace: namespace is a byte slice of length 33 where the first byte is the
+  namespaceVersion and the subsequent 32 bytes are the namespaceId.
+- blob_sizes: sizes of each blob in bytes.
+- share_commitments is a list of share commitments (one per blob).
+- share_versions are the versions of the share format that the blobs associated
+  with this message should use when included in a block. The share_versions
+  specified must match the share_versions used to generate the share_commitment
+  in this message. See
+  [ADR014](../../docs/architecture/adr-014-versioned-namespaces.md) for more
+  details on how this effects the share encoding and when it is updated.
+
+Note that while the shares version in each protobuf encoded PFB are uint32s, the
+internal represantation of shares versions is always uint8s. This is because
+protobuf doesn't support uint8s.
+
+### Generating the `ShareCommitment`
+
+The share commitment is the commitment to share encoded blobs. It can be used
+for cheap inclusion checks for some data by light clients. More information and
+rational can be found in the [data square layout
+specs](../../specs/src/specs/data_square_layout.md).
+
+1. Split the blob into shares of size [`shareSize`](../../specs/src/specs/data_structures.md#consensus-parameters)
+1. Determine the
+   [`SubtreeWidth`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/pkg/shares/non_interactive_defaults.go#L94-L116)
+   by dividing the length in shares by the `SubtreeRootThreshold`.
+1. Generate each subtree root by diving the blob shares into `SubtreeWidth`
+   sized sets, then take the binary [namespaced merkle tree
+   (NMT)](https://github.com/celestiaorg/nmt/blob/v0.16.0/docs/spec/nmt.md) root
+   of each set of shares.
+1. Calculate the final share commitment by taking the merkle root (note: not an
+   NMT, just a normal binary merkle root) of the subtree roots from the previous
+   step.
+
+See
+[`CreateCommitment`](https://github.com/celestiaorg/celestia-app/blob/v1.0.0-rc2/x/blob/types/payforblob.go#L169-L236)
+for an implementation. See [data square
+layout](../../specs/src/specs/data_square_layout.md) and
+[ADR013](../../docs/architecture/adr-013-non-interactive-default-rules-for-zero-padding.md)
+for details on the rational of the square layout.
+
+## Validity Rules
+
+In order for a proposal block to be considered valid, each `BlobTx`, and thus
+each PFB, to be included in a block must follow a set of validity rules.
+
+1. Signatures: All blob transactions must have valid signatures. This is
```

Comment:
[Question for my own understanding] When the term "proposal block" is used in the context of "In order for a proposal block to be considered valid," does it refer to a block proposed during the consensus (and yet to be voted and committed) or one that has received the majority vote (hence committed)? I'm curious about when signatures and nonces are verified – during the voting phase or at the time of block execution (which I think corresponds to the processProposal). If signature verification is required during the voting process, then we would need to track the state changes resulting from proposals (but not committed blocks) in order to do so.

Reply:
> does it refer to a block proposed during the consensus (and yet to be voted and committed) or one that has received the majority vote (hence committed)?

both! we call ProcessProposal on proposals beofore voting, and while consensus nodes sync the chain

signatures, and therefore their nonces, are verified before the block is actually applied to the state in both cases. State can then be tracked rather easily since we already have the state from the previous block, which is all that is needed.